lib/cancan/model_adapters/active_record_5_adapter.rb

```diff
@@ -21,17 +21,57 @@ def self.matches_condition?(subject, name, value)
 
       private
 
+      delegate :connection, :quoted_primary_key, to: :@model_class
+      delegate :quote_table_name, to: :connection
+
       def build_joins_relation(relation, *where_conditions)
         case CanCan.accessible_by_strategy
         when :subquery
-          inner = @model_class.unscoped do
-            @model_class.left_joins(joins).where(*where_conditions)
-          end
-          @model_class.where(@model_class.primary_key => inner)
-
+          build_joins_relation_subquery(where_conditions)
         when :left_join
           relation.left_joins(joins).distinct
+        when :double_exist_subquery
+          build_joins_relation_double_exist_subquery(where_conditions)
+        end
+      end
+
+      def build_joins_relation_subquery(where_conditions)
+        inner = @model_class.unscoped do
+          @model_class.left_joins(joins).where(*where_conditions)
         end
+        @model_class.where(@model_class.primary_key => inner)
+      end
+
+      def build_joins_relation_double_exist_subquery(where_conditions)
+        @model_class.where("EXISTS (#{double_exists_query_sql(where_conditions)})")
+      end
+
+      def double_exists_inner_query(where_conditions)
```

Comment:
I am not sure the code has gotten less abstract having to use this many helper methods, but I couldn't otherwise find a way around the Rubocop complexity offences :'-(

Reply:
I honestly might get rid of rubocop and switch to standard soon :)

Reply:
Personally I feel that Rubocop has forced me to make the code more abstract than it had to be here :-)